### ISSUE TITLE
Shimmie API

### ISSUE THREAD

**Issue by sanmadjack** [shish/shimmie2#788]
Shimmie could use a standard API that users can connect to, but more importantly can be used for more advanced in-page interactions such as autocomplete. Current autocomplete uses its own endpoint to get tag info, but it would be better for it to use an endpoint that is part of a standardized interface.

What are the downfalls of the current APIs?

1. They're specific to the application that connects to them.
2. The JSON API "leaks" data.
3. Vast majority of shimmie functionality not appropriate for the implementation, or is just unimplemented.

What do we need out of an API?

1. Authentication. Currently the user API key works for this, but more advanced authentication can be implemented.
2. Comprehensive coverage. The system has lots of pieces. How do we want to organize it all?
3. Profit?

**Comment by shish** [comment]
Brain-dumping a stream of thought, apologies if ideas at the start contradict ideas at the end :D

General formats I've worked with in the past:

- Have a whole bunch of independent HTTP endpoints which all return JSON in a vaguely-consistent format
  - each extension listens for `PageRequestEvent(/api/$extension/*)` and responds in its own way

- Have a single endpoint launching `Events` and returning data in a standard format
  - Right now the two entry points to the system are `PageRequestEvent` (outputs HTML) and `CommandEvent` (outputs plain text); I feel like `ApiRequestEvent` (outputs JSON) is a natural extension?
  - This feels like it maps nicely to shimmie's architecture with a whole bunch of independent extensions
  - "Wherever current extensions support PageRequestEvent, they should also try to support ApiRequestEvent" seems like a simple and straightforward approach to getting complete coverage

- GraphQL
  - This has been really nice from the client side, pretty much allowing the client to specify what data they need, instead of requiring the server to build APIs for every client-side use-case
  - I've never tried implementing the server side, my gut feeling is that it might be a nightmare
  - In particular, does it support adding and removing API features dynamically (eg when you enable the "ratings" extension, we want `Post` objects to include a `ratings` field)
  - Is this going to end up being basically a whole separate server-side application, which uses Shimmie's database but has nothing else in common with the rest of the codebase?
  - What would an extension's code look like in this world?

My gut feeling is to go with the second one, as that feels like it fits in very nicely with the existing architecture, and would be the smallest amount of work for each extension to be supported; but from a client-side perspective it's not quite as nice as having an open-ended "query for whatever information you need" endpoint.

JSON-RPC looks like it might be a useful format which also maps nicely to `ApiRequestEvent` / `ApiResponse` classes.

I'm imagining you could `POST` a batch of requests to `/api/json-rpc` (authenticated via HTTP headers?):

```
[
    {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
    {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
    {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
    {"jsonrpc": "2.0", "method": "get_data", "id": "9"} 
]
```

Which then fires off a set of `ApiRequestEvent`s, then collates the results, and returns

```
[
    {"jsonrpc": "2.0", "result": 7, "id": "1"},
    {"jsonrpc": "2.0", "result": 19, "id": "2"},
    {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
]
```

So that seems like a nice standard API that would also be nice to implement behind the scenes... but does it do the things that a hypothetical client app would need?

**Comment by sanmadjack** [comment]
I've done some reading into GraphQL, I think it might work well. I'm going to make a test implementation to see what integrating looks like.

**Comment by shish** [comment]
Cool beans :)

I am tempted to make a pull request for a json-rpc API just because I can see all the code for it in my head already, and it'd be an interesting comparison (and then abandon it if GraphQL turns out to be cleaner to implement than I thought, as I'd rather support one good API than five half-assed ones :P)

**Comment by sanmadjack** [comment]
I'll be honest, I see a lot of potential for this GraphQL implementation, but it's not a bad idea to explore multiple options while we're still starting out so we don't miss out on what could be the better implementation. I've tinkered with json-rpc for my handbrake web frontend, and it does work fairly well. GraphQL's advantage seems to be that it can do some pretty crazy custom querying without having to explicitly define it on the server, just defining how things work at each layer of the data. For instance, I'm currently able to query a post, get the author, everything they've posted, and then get those authors (which are the same author of course) and continuing recursively as far as I want. I'm only starting to scratch the surface of write operations, but it looks similarly dynamic.

I've already got reading posts, tags, and users working (not 100%, but enough to understand how things will fit together).  I want to get some more samples of write operations implemented before I feel like it's enough of a vertical slice for me to send the pull for review. I also want to implement a sample of pagination support.
